pydantic/config.py

```diff
@@ -34,185 +35,175 @@ class Extra(str, Enum):
     validate_assignment: bool
-    error_msg_templates: Dict[str, str]
-    arbitrary_types_allowed: bool
-    orm_mode: bool
+    arbitrary_types_allowed: bool  # TODO default True, or remove
+    undefined_types_warning: bool  # TODO review docs
+    from_attributes: bool
     alias_generator: Optional[Callable[[str], str]]
-    keep_untouched: Tuple[type, ...]
-    schema_extra: Union[Dict[str, object], 'SchemaExtraCallable']
-    json_loads: Callable[[str], object]
-    json_dumps: Callable[..., Any]
-    json_encoders: Dict[Type[object], Callable[..., Any]]
-    underscore_attrs_are_private: bool
+    keep_untouched: Tuple[type, ...]  # TODO remove??
+    schema_extra: Union[Dict[str, Any], 'SchemaExtraCallable']  # TODO remove, new model method
+    json_loads: Callable[[str], Any]  # TODO decide
+    json_dumps: Callable[..., str]  # TODO decide
+    json_encoders: Dict[Union[Type[Any], str, ForwardRef], Callable[..., Any]]  # TODO decide
     allow_inf_nan: bool
-    copy_on_model_validation: Literal['none', 'deep', 'shallow']
-    post_init_call: Literal['before_validation', 'after_validation']
 
+    strict: bool
+
+    # whether inherited models as fields should be reconstructed as base model,
+    # and whether such a copy should be shallow or deep
+    copy_on_model_validation: Literal['none', 'deep', 'shallow']  # TODO remove???
+
+    # whether dataclass `__post_init__` should be run before or after validation
+    post_init_call: Literal['before_validation', 'after_validation']  # TODO remove
+
+
+config_keys = set(_ConfigDict.__annotations__.keys())
 
-config_keys = set(ConfigDict.__annotations__.keys())
+
+if TYPE_CHECKING:
+
+    class ConfigDict(_ConfigDict):
+        ...
+
+else:
+
+    class ConfigDict(dict):
+        def __missing__(self, key):
+            return _default_config[key]
+
+
+_default_config = ConfigDict(
+    title=None,
+    str_to_lower=False,
+    str_to_upper=False,
+    str_strip_whitespace=False,
+    str_min_length=0,
+    str_max_length=None,
+    extra=Extra.ignore,
+    frozen=False,
+    populate_by_name=False,
+    use_enum_values=False,
+    validate_assignment=False,
+    arbitrary_types_allowed=False,
+    undefined_types_warning=True,
+    from_attributes=False,
+    alias_generator=None,
+    keep_untouched=(),
+    json_loads=json.loads,
+    json_dumps=json.dumps,
+    json_encoders={},
+    allow_inf_nan=True,
+    strict=False,
+    copy_on_model_validation='shallow',
+    post_init_call='before_validation',
+)
 
 
 class BaseConfig:
     title: Optional[str] = None
-    anystr_lower: bool = False  # TODO rename to str_to_lower
-    anystr_upper: bool = False  # TODO rename to str_to_upper
-    anystr_strip_whitespace: bool = False  # TODO rename to str_strip_whitespace
-    min_anystr_length: int = 0  # TODO rename to str_min_length
-    max_anystr_length: Optional[int] = None  # TODO rename to str_max_length
-    validate_all: bool = False  # TODO remove
+    str_to_lower: bool = False
```

Comment:
I wonder if you can remove all these attributes from `BaseClass`, perhaps add a `__getattr__` method to with a warning?

Reply:
You mean all attributes or just the ones which were renamed, so that you get a warning if you use f.i. `anystr_to_lower` instead of `str_to_lower` ? If we do it for every attribute then the user would get duplicate warnings when using BaseClass instead of ConfigDict and when using an attribute of the class... And you lose intellisense.
But if you wish i can change it.

Reply:
I was suggesting removing all, usage is not recommended after all. But totally up to you.

Reply:
I was not able to get this working, because if you use `BaseConfig.str_to_lower` `__getattr__`  won't be called and you would probably need a metaclass which implements this method and i don't know if it's worth to do that and what implication this would have...

Reply:
Ye, I thought we might need a custom metaclass. I'll look next week.

@dmontagu WDYT? On this specifically and the whole "config is a dict" thing.